DHFPROD-5682:5788: Retrieve and apply keystore from cloud services

akshaysonvane opened the pull request:
### Description
Introduced two new spring profiles specific to Azure and AWS which when activated will try to fetch the certificate artifacts from Azure/AWS services using the configured properties.

Following properties are needed for configuration:

AWS properties
```
spring.profiles.active=aws
aws.s3.bucketName=
aws.s3.keyName=
```

Azure properties
```
spring.profiles.active=azure
azure.keyvault.uri=
azure.storage.endpoint=
azure.storage.containerName=
azure.storage.blobName=
```

#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [N/A] Added Tests
  

- ##### Reviewer:

- [N/A] Reviewed Tests

rjrudin commented on marklogic-data-hub-central/build.gradle:
```diff
@@ -68,6 +68,15 @@ dependencies {
         exclude group: "com.squareup.okhttp3", module: "logging-interceptor"
     }
+
+    // Excluding okhttp3 dependency as it bumps down the dependency from 4.4.0 to 3.x
```
Very helpful comment, thanks

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/keystore/KeyStoreInit.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.io.InputStream;
+
+public interface KeyStoreInit {
```
I don't think you need this, as there's no client for this interface.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/keystore/AzureKeyStoreInit.java:
```diff
@@ -0,0 +1,59 @@
+    @Value("${azure.storage.endpoint}")
+    private String endpoint;
+
+    @Value("${azure.storage.container.name}")
```
Just do "containerName" and "blobName", unless we think we're going to have other properties on objects named "container" and "blob".

akshaysonvane replied:
This is to maintain naming consistency with the spring property (`azure.keyvault.uri`) added by `azure-keyvault-secrets-spring-boot-starter` dependency.
Plus it allows us to easily identify the cloud platform and the cloud service that the property is being used for.

rjrudin replied:
The question is whether the "container" should be an object for property-naming purposes. If the property name on "storage" is "container name", then we should use "containerName" - we don't want to use a period as a substitute for a space. But if "container" is essentially a structured property of "storage" and "name" is just one property of "container", then "container.name" is correct. Same logic applies for "bucket".

Basically - we want to use a period to indicate that something is conceptually an object, not as a substitute for a space.

akshaysonvane replied:
Fair enough. I'll change this to `azure.storage.containerName` and make similar updates to other properties.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/keystore/AWSKeyStoreInit.java:
```diff
@@ -0,0 +1,56 @@
+@Profile("aws")
+public class AWSKeyStoreInit extends AbstractKeyStoreInit {
+
+    @Value("${aws.s3.bucket.name}")
```
Just use "bucketName" and "keyName" unless we're certain that we'll have other properties on objects named "bucket" and "key".

akshaysonvane replied:
Used the same naming convention that we used for Azure i.e `cloud-platform.cloud-service.property-name`

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/keystore/AzureKeyStoreInit.java:
```diff
@@ -0,0 +1,59 @@
+@Profile("azure")
+public class AzureKeyStoreInit extends AbstractKeyStoreInit {
+
+    @Value("${key-store-password}")
```
Shouldn't this be e.g. "azure.storage.keyStorePassword"? It's the password that's specific to the keystore for Azure, right?

akshaysonvane replied:
Azure KeyVault stores secrets as a key-value pair.
Here the key is `key-store-password` and the value is the actual password.
Azure does not allow `.` in key names so this is what I've used. 

Also, the key name is kept consistent across both Azure and AWS services.
For AWS we'll fetch value for the key `key-store-password` from the AWS parameter store.

rjrudin replied:
Got it, makes sense.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/keystore/AWSKeyStoreInit.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.io.InputStream;
+
+@Component
```
I believe this and the Azure class should really be Configuration classes, as they need to configure the app when Spring is starting up. We're likely to have other AWS/Azure-specific beans, so I think this should be AWSConfiguration and the other class should be AzureConfiguration. And they should be in the com.marklogic.hub.central.cloud.(aws|azure) package.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/keystore/AbstractKeyStoreInit.java:
```diff
@@ -0,0 +1,66 @@
+        return serverProperties;
+    }
+
+    @Bean
```
"@Bean" is not inheritable, so I'm curious how these would work.

I think the better approach is to declare both of these Bean methods on AWSConfiguration and AzureConfiguration, and then move this reusable logic into a CloudProviderUtil class in the com.marklogic.hub.central.cloud package (with the AWS class in cloud.aws and the Azure class in cloud.azure). 

Generally, for reusing functionality, it's better to use composition instead of inheritance. Inheritance can be handy for reusing state - i.e. fields defined in a parent class. 

CloudProviderUtil would then have two methods - buildServerProperties(keyStorePassword) and configureSslStoreProvider(tomcatServletWebServerFactory, keyStorePassword, keyStoreFile).

akshaysonvane commented:
- Updated the property names.
- Moved the AWS and Azure configurations classes under com.marklogic.hub.central.cloud.aws/azure package. 
- Moved the common code to CloudProviderUtil class under com.marklogic.hub.central.cloud package.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/cloud/CloudProviderUtil.java:
```diff
@@ -0,0 +1,64 @@
+ *  limitations under the License.
+ *
+ */
+package com.marklogic.hub.central.cloud;
```
I gave bad advice on this class package. It doesn't have anything to do with cloud providers, but it is web-specific and SSL-specific. So let's call this SslUtil and toss it com.marklogic.hub.central.web (and at some point, we probably ought to move our controllers there, but that's a separate matter).

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/cloud/aws/AWSConfiguration.java:
```diff
@@ -0,0 +1,71 @@
+        return CloudProviderUtil.configureSslStoreProvider(getKeyStorePassword(), getKeyStoreFile());
+    }
+
+    private String getKeyStorePassword() {
```
Let's add logging to both of these private methods. Nothing special - just at the start of each method, log "Retrieving keystore password", and then at the end, "Retrieved keystore password". And call this method "retrieveKeyStorePassword" to make it clear that this is not a getter (I like for "get" to imply that it's only returning the value of a local variable). 

Same for the file - "Retrieving keystore file" and then "Retrieved keystore file". 

The two benefits are - easy confirmation that the system is using the correct profile, and we can also see how long the calls are taking too.

akshaysonvane commented:
- Moved the common code to SslUtil class under com.marklogic.hub.central.web package.
- Added log statements before and after artifact retrieval.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/cloud/aws/AWSConfiguration.java:
```diff
@@ -0,0 +1,89 @@
+@Profile("aws")
+public class AWSConfiguration {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AWSConfiguration.class);
```
For consistency, just extend LoggingObject, which defines this for you.

akshaysonvane replied:
I was going to extend `LoggingObject` but didn't since we are basically resuing functionality and so did not prefer to inherit.

rjrudin replied:
Okay - can you use the variable name "logger" then? Just wanted to avoid future questions of "Why is it logger in some places but LOGGER here?"

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/cloud/azure/AzureConfiguration.java:
```diff
@@ -0,0 +1,75 @@
+@Profile("azure")
+public class AzureConfiguration {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AzureConfiguration.class);
```
Same, just extend LoggingObject.

akshaysonvane commented:
- Changed variable name to `logger`
- Did not find a simple way to test out the SslUtil class since tomcat configuration happens very early on in the spring boot lifecycle and I can't seem to configure it again later.
